clippy_utils/src/sugg.rs

```diff
@@ -871,7 +871,11 @@ impl<'tcx> DerefDelegate<'_, 'tcx> {
             },
             ExprKind::Call(func, call_args) => {
                 let typ = self.cx.typeck_results().expr_ty(func);
-                (call_args, typ.fn_sig(self.cx.tcx).skip_binder().inputs())
+                match typ.kind() {
+                    ty::Closure(_, substs) => (call_args, substs.as_closure().sig().skip_binder().inputs()),
+                    ty::FnDef(..) | ty::FnPtr(_) => (call_args, typ.fn_sig(self.cx.tcx).skip_binder().inputs()),
+                    _ => return false,
+                }
             },
             _ => return false,
         };
```

Comment:
It would look more like
```rust
let ty = match parent_expr.kind {
    ExprKind::MethodCall(_, call_args, _) => {
        if let Some(sig) = .. {
            call_args.iter().position(|arg| arg.hir_id == cmt_hir_id).map(|i| sig.inputs()[i])
        } else {
            return false;
        }
    }
    ExprKind::Call(func, args) => {
        if let Some(sig) = expr_sig(func) {
            args.iter().position(|arg| arg.hir_id == cmt_hir_id).and_then(|i| sig.input(i))
        } else {
            return false;
        }
    }
    _ => return false;
}
```

Reply:
Like this?

And I do stand corrected, I now see that there is test coverage for the `ExprKind::MethodCall` branch. I'm not sure how I was missing them previously.

Reply:
That looks right. Should probably add a test for both an `impl Fn()` type and a `dyn Fn()` type as the callee.